### DISCUSSION TITLE
Custom Extension: Set extension value via CMP

### DISCUSSION THREAD

**Discussion by frankschierle** [Keyfactor/ejbca-ce#465] [Q&A] [answered]
Hi there,

we are using EJBCAs CMP endpoint to issue certificates. No we have to add a new custom X509 extension to our certificates. The value of the extension is not static and should be provided in the certificate request.

In EJBCA we added an new custom extension of type `DERUTF8STRING` in` System configuration -> Custom Certificate Extensions`:
![image](https://github.com/Keyfactor/ejbca-ce/assets/21262852/1202b189-e594-412f-8fd2-2c609e1c2003)
(OID 1.2.3.4 is just used for testing puposes in my local environment. A proper OID will be used later)

The certificate profile is configured to include the new extension and `Custom certificate extension data` is enabled in the end entity profile.

We added the extension to the Cert Template of our CMP certificate request. However, the issued certificate always includes the default value, defined in the screenshot above. If i specify the `Custom certificate extension data` at the end entity manually and issue a certificate via RA web, the correct value (as specified in `Custom certificate extension data`) is included in the certificate.

I had a look in source code and it looks like, EJBCA only respects the `KeyUsage` extension. All other extensions will be ignored:
https://github.com/Keyfactor/ejbca-ce/blob/0966b4269d44e2aedc70731829ad59603774358b/modules/cesecore-ejb/src/org/cesecore/certificates/certificate/CertificateCreateSessionBean.java#L211-L225

How can we specify the value of our custom extension in a CMP certificate request? Or is it not possible?

Thank you for your help!

**Comment by primetomas** [comment, marked as answer]
If you want to send in the extension in the request, you do not need to configure custom extensions. You would use the "Allow Extension Override" function in the certificate profiles. That function specifies what extensions will be grabbed directly from the CSR.

**Reply by frankschierle** [reply]
Thank you very much! It works perfectly!

I didn't think it would be that simple.

**Reply by primetomas** [reply]
Oh, PKI is very simple ;-)
